Sources/Features/Client/BacktraceOomWatcher.swift

```diff
@@ -0,0 +1,148 @@
+        
+        // set status file url
+        if(BacktraceOomWatcher.oomFilePath == nil) {
+            // database path will point out sqlite database path
+            // oom status should exist in the same directory where database exists.
+            BacktraceOomWatcher.oomFilePath  = self.repository.url.deletingLastPathComponent().absoluteURL
+                .appendingPathComponent("BacktraceOOMState")
+        }
+        self.statusFilePath = BacktraceOomWatcher.oomFilePath!
+    }
+    
+    internal static func clean() {
+        if((BacktraceOomWatcher.oomFilePath) != nil) {
+            try! FileManager.default.removeItem(at: BacktraceOomWatcher.oomFilePath!)
+        }
+    }
+    
+    public func start() {
+        sendPendingOomReports()
+        // override previous application state after reading all information
+        saveState()
+    }
+    
+    internal static func getStatusFilePath() -> URL? {
+        // oom status file is available in application cache dir - the same dir
+        // where we store attributes
+        
+        guard let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            return nil
+        }
+        return cacheDir.appendingPathComponent("BacktraceOomState.plist")
+    }
+    
+    private func sendPendingOomReports() {
+        
+        // if oom state file doesn't exist it means that we deleted it to
+        // prevent sending false oom crashes
+        if(!FileManager.default.fileExists(atPath: statusFilePath.path)){
+            return
+        }
+        
+        guard let backtraceOOMState = NSDictionary(contentsOf: statusFilePath) else {
+            return
+        }
+        
+        // no low memory warning
+        if(backtraceOOMState["resource"] == nil) {
+            return
+        }
+        
+        // check if debugger was enabled
+        if(backtraceOOMState["debugger"] as! Bool == true) {
+            // detected system update
+            return
+        }
+        // check system update
+        if(backtraceOOMState["version"] as! String !=  UIDevice.current.systemVersion as String) {
+            // detected system update
+            return
+        }
+        
+        // check application update
+        if(backtraceOOMState["appVersion"] as! String != Bundle.main.object(forInfoDictionaryKey: kCFBundleVersionKey as String) as! String ) {
+            // detected app update
+            return
+        }
+        
+        var attributes = backtraceOOMState["resource.attributes"] as! Attributes
+        attributes["error.message"] = "Out of memory detected."
+        attributes["error.type"] = "Low Memory"
+        attributes["state"] = backtraceOOMState["state"]
+        
+        // ok - we detected oom and we should report it
+        guard let report = try? BacktraceReport(report: backtraceOOMState["resource"] as! Data, attributes: attributes,   attachmentPaths: []) else {
```

Comment:
Very cool! Is there some way we could differentiate BOOM vs. FOOM here? I see we have hooks into background/foreground transition events in this pull request as well